substrate/client/src/notifications.rs

```diff
@@ -0,0 +1,267 @@
+
+//! Storage notifications
+
+use std::{
+	collections::{HashSet, HashMap},
+	sync::Arc,
+};
+
+use fnv::{FnvHashSet, FnvHashMap};
+use futures::sync::mpsc;
+use primitives::storage::{StorageKey, StorageData};
+use runtime_primitives::traits::Block as BlockT;
+
+/// Storage change set
+#[derive(Debug)]
+pub struct StorageChangeSet {
+	changes: Arc<Vec<(StorageKey, Option<StorageData>)>>,
+	filter: Option<HashSet<StorageKey>>,
+}
+
+impl StorageChangeSet {
+	/// Convert the change set into iterator over storage items.
+	pub fn iter<'a>(&'a self) -> impl Iterator<Item=&'a (StorageKey, Option<StorageData>)> + 'a {
+		self.changes
+			.iter()
+			.filter(move |&(key, _)| match self.filter {
+				Some(ref filter) => filter.contains(key),
+				None => true,
+			})
+	}
+}
+
+/// Type that implements `futures::Stream` of storage change events.
+pub type StorageEventStream<H> = mpsc::UnboundedReceiver<(H, StorageChangeSet)>;
+
+type SubscriberId = u64;
+
+/// Manages storage listeners.
+#[derive(Debug)]
+pub struct StorageNotifications<Block: BlockT> {
+	next_id: SubscriberId,
+	wildcard_listeners: FnvHashSet<SubscriberId>,
+	listeners: HashMap<StorageKey, FnvHashSet<SubscriberId>>,
+	sinks: FnvHashMap<SubscriberId, (
+		mpsc::UnboundedSender<(Block::Hash, StorageChangeSet)>,
+		Option<HashSet<StorageKey>>,
+	)>,
+}
+
+impl<Block: BlockT> Default for StorageNotifications<Block> {
+	fn default() -> Self {
+		StorageNotifications {
+			next_id: Default::default(),
+			wildcard_listeners: Default::default(),
+			listeners: Default::default(),
+			sinks: Default::default(),
+		}
+	}
+}
+
+impl<Block: BlockT> StorageNotifications<Block> {
+	/// Trigger notification to all listeners.
+	///
+	/// Note the changes are going to be filtered by listener's filter key.
+	/// In fact no event might be sent if clients are not interested in the changes.
+	pub fn trigger(&mut self, hash: &Block::Hash, changeset: impl Iterator<Item=(Vec<u8>, Option<Vec<u8>>)>) {
+		let has_wildcard = !self.wildcard_listeners.is_empty();
+
+		// early exit if no listeners
+		if !has_wildcard && self.listeners.is_empty() {
+			return;
+		}
+
+		let mut subscribers = self.wildcard_listeners.clone();
+		let mut changes = Vec::new();
+
+		// Collect subscribers and changes
+		for (k, v) in changeset {
+			let k = StorageKey(k);
+			let listeners = self.listeners.get(&k);
+
+			if let Some(ref listeners) = listeners {
+				subscribers.extend(listeners.iter());
+			}
+
+			if has_wildcard || listeners.is_some() {
+				changes.push((k, v.map(StorageData)));
+			}
+		}
+
+		let changes = Arc::new(changes);
+		// Trigger the events
+		for subscriber in subscribers {
+			let remove_subscriber = match self.sinks.get(&subscriber) {
+				Some(&(ref sink, ref filter)) => {
+					sink.unbounded_send((hash.clone(), StorageChangeSet {
+						changes: changes.clone(),
+						filter: filter.clone(),
+					})).is_err()
+				},
+				None => false,
```

Comment:
If I read this right, if the subscriber is gone when we get here, then we assume they've been removed properly already, so we're returning `false` to avoid calling `remove_subscriber()` again for them? It's a bit unclear to me how they can still be in the `subscribers` collection though, can you elaborate on that?

Reply:
Indeed, I could actually `.expect()` here, since if the structure is consistent the `subscribers` should always be in `self.sinks`. The check here is superfluous.
I could refactor to:
```rust
let &(ref sink, ref filter) = self.sinks.get(&subscriber).expect("subscribers returned from self.listeners are always in self.sinks; qed");
let result = sink.unbounded_send((hash.clone(), StorageChangeSet {
    changes: changes.clone(),
    filter: filter.clone(),
 }));
if result.is_err() {
  self.remove_subscriber(subscriber);
}
```
or 
```rust
if let Some(&(ref sink, ref filter)) = match self.sinks.get(&subscriber) {
   let result = sink.unbounded_send((hash.clone(), StorageChangeSet {
      changes: changes.clone(),
      filter: filter.clone(),
   }));
   if result.is_err() {
     self.remove_subscriber(subscriber);
   }
}
```

Which one do you prefer?

Reply:
Oh, actually I can't since `.get()` borrows immutably, so `remove_subscriber` has to be outside of the scope.